File: libs/nol_nole.py
import os, re, unicodedata, time
from enum import Enum, auto
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Set, Union

class NolError(Exception): pass

class Tok(Enum):
    EOF = auto(); Identifier = auto(); String = auto(); Number = auto(); Date = auto()
    True_ = auto(); False_ = auto(); Null = auto(); Colon = auto(); Comma = auto()
    LBracket = auto(); RBracket = auto(); LBrace = auto(); RBrace = auto(); Newline = auto()
    Dot = auto(); Star = auto(); LShift = auto(); Ampersand = auto(); LArrow = auto(); RArrow = auto()

@dataclass
class Token:
    type: Tok; text: str; line: int; col: int

class Lexer:
    def __init__(self, input_str: str):
        self.input = input_str; self.pos = 0; self.line = 1; self.col = 1
        if len(input_str) > 10 * 1024 * 1024: raise NolError("Document size limit exceeded")
        if '\0' in input_str: raise NolError("Null bytes are prohibited")
        for c in input_str:
            if 0xD800 <= ord(c) <= 0xDFFF: raise NolError("Surrogate code points are prohibited")

    def peek(self, n=0): return self.input[self.pos + n] if self.pos + n < len(self.input) else '\0'
    def advance(self):
        c = self.peek(); self.pos += 1
        if c == '\n': self.line += 1; self.col = 1
        else: self.col += 1
        return c
    def next_token(self) -> Token:
        while True:
            c = self.peek()
            if c == '\0': return Token(Tok.EOF, "", self.line, self.col)
            if c.isspace():
                if c == '\n': t = Token(Tok.Newline, "\n", self.line, self.col); self.advance(); return t
                self.advance(); continue
            if c == '#':
                self.advance()
                if self.peek() == '#':
                    self.advance()
                    while self.peek() != '\0':
                        if self.peek() == '#' and self.peek(1) == '#': self.advance(); self.advance(); break
                        self.advance()
                else:
                    while self.peek() != '\0' and self.peek() != '\n': self.advance()
                continue
            if c == '"' or c == "'": return self.scan_string()
            if c.isdigit() or (c == '-' and self.peek(1).isdigit()): return self.scan_number_or_date()
            if c.isalpha() or c == '_': return self.scan_identifier()
            if c == ':': self.advance(); return Token(Tok.Colon, ":", self.line, self.col-1)
            if c == ',': self.advance(); return Token(Tok.Comma, ",", self.line, self.col-1)
            if c == '[': self.advance(); return Token(Tok.LBracket, "[", self.line, self.col-1)
            if c == ']': self.advance(); return Token(Tok.RBracket, "]", self.line, self.col-1)
            if c == '{': self.advance(); return Token(Tok.LBrace, "{", self.line, self.col-1)
            if c == '}': self.advance(); return Token(Tok.RBrace, "}", self.line, self.col-1)
            if c == '.': self.advance(); return Token(Tok.Dot, ".", self.line, self.col-1)
            if c == '*': self.advance(); return Token(Tok.Star, "*", self.line, self.col-1)
            if c == '&': self.advance(); return Token(Tok.Ampersand, "&", self.line, self.col-1)
            if c == '<':
                self.advance()
                if self.peek() == '<': self.advance(); return Token(Tok.LShift, "<<", self.line, self.col-2)
                return Token(Tok.LArrow, "<", self.line, self.col-1)
            if c == '>': self.advance(); return Token(Tok.RArrow, ">", self.line, self.col-1)
            raise NolError(f"Unexpected character: {c} at {self.line}:{self.col}")
    def scan_string(self) -> Token:
        quote = self.advance(); is_triple = (self.peek() == quote and self.peek(1) == quote)
        if is_triple: self.advance(); self.advance()
        res = []
        while True:
            c = self.peek()
            if c == '\0': raise NolError("Unterminated string")
            if is_triple:
                if c == quote and self.peek(1) == quote and self.peek(2) == quote: self.advance(); self.advance(); self.advance(); break
            else:
                if c == quote: self.advance(); break
                if c == '\n': raise NolError("Newline in string")
            if c == '\\' and quote == '"':
                self.advance(); e = self.advance()
                if e == '"': res.append('"')
                elif e == "'": res.append("'")
                elif e == '\\': res.append('\\')
                elif e == 'n': res.append('\n')
                elif e == 'r': res.append('\r')
                elif e == 't': res.append('\t')
                elif e in ('u', 'U'):
                    n = 4 if e == 'u' else 8; u = "".join(self.advance() for _ in range(n))
                    cp = int(u, 16)
                    if cp == 0: raise NolError("Null prohibited")
                    if 0xD800 <= cp <= 0xDFFF: raise NolError("Surrogate prohibited")
                    res.append(chr(cp))
                else: raise NolError(f"Invalid escape: \\{e}")
            else: res.append(self.advance())
            if len(res) > 1024 * 1024: raise NolError("String length limit exceeded")
        return Token(Tok.String, "".join(res), self.line, self.col)
    def scan_number_or_date(self) -> Token:
        res = []
        while self.peek().isalnum() or self.peek() in "-._:T+": res.append(self.advance())
        s = "".join(res)
        if re.match(r'^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(Z|[+-]\d{2}:\d{2})?)?$', s): return Token(Tok.Date, s, self.line, self.col)
        return Token(Tok.Number, s, self.line, self.col)
    def scan_identifier(self) -> Token:
        res = []
        while self.peek().isalnum() or self.peek() in "_-": res.append(self.advance())
        s = "".join(res)
        kw = {"true": Tok.True_, "false": Tok.False_, "null": Tok.Null}
        return Token(kw.get(s, Tok.Identifier), s, self.line, self.col)

class Parser:
    def __init__(self, input_str: str, is_nole: bool = False, env_allowlist: List[str] = None):
        self.lexer = Lexer(input_str); self.cur = self.lexer.next_token(); self.is_nole = is_nole
        self.app_env_allowlist = set(env_allowlist or []); self.doc_env_allowlist = set()
        self.anchors = {}; self.root = {}; self.depth = 0; self.start_time = time.time()
    def error(self, msg): raise NolError(f"{msg} at {self.cur.line}:{self.cur.col}")
    def next(self): self.cur = self.lexer.next_token()
    def skip_newlines(self):
        while self.cur.type == Tok.Newline: self.next()
    def parse(self) -> Dict[str, Any]:
        while self.cur.type != Tok.EOF:
            self.skip_newlines()
            if self.cur.type == Tok.EOF: break
            if self.cur.type == Tok.LBracket: self.parse_section()
            elif self.is_key_token(): self.parse_pair(self.root)
            else: self.error(f"Unexpected token: {self.cur.type}")
            if time.time() - self.start_time > 1.0: raise NolError("Parse timeout")
        if self.is_nole:
            self.collect_env_allowlist(self.root)
            self.root = self.resolve_merges(self.root)
            self.root = self.resolve_env(self.root)
            self.root = self.resolve_interpolations(self.root)
            self.root = self.resolve_coercions(self.root)
        return self.root
    def collect_env_allowlist(self, obj):
        if isinstance(obj, dict):
            if "_env" in obj and isinstance(obj["_env"], dict):
                allowed = obj["_env"].get("allowed")
                if isinstance(allowed, list):
                    for a in allowed: self.doc_env_allowlist.add(str(a))
            for v in obj.values(): self.collect_env_allowlist(v)
        elif isinstance(obj, list):
            for v in obj: self.collect_env_allowlist(v)
    def is_key_token(self): return self.cur.type in (Tok.Identifier, Tok.String, Tok.LShift)
    def parse_section(self):
        self.next()
        path = []; is_array = False
        while True:
            if self.cur.type == Tok.Star: is_array = True; self.next()
            elif self.cur.type in (Tok.Identifier, Tok.String): path.append(unicodedata.normalize('NFC', self.cur.text)); self.next()
            else: self.error("Expected section name")
            if self.cur.type == Tok.Dot: self.next(); continue
            if self.cur.type == Tok.RBracket: self.next(); break
            self.error("Expected . or ]")
        target = self.root
        for i, part in enumerate(path):
            if part in target:
                if not isinstance(target[part], dict) and not (is_array and i == len(path)-1 and isinstance(target[part], list)): self.error(f"Collision at {part}")
            else: target[part] = [] if (is_array and i == len(path)-1) else {}
            if not (is_array and i == len(path)-1): target = target[part]
        if is_array:
            arr = target[path[-1]]; new_obj = {}; arr.append(new_obj); target = new_obj
        self.skip_newlines()
        while self.is_key_token(): self.parse_pair(target); self.skip_newlines()
    def parse_pair(self, obj):
        key = "<<" if self.cur.type == Tok.LShift else unicodedata.normalize('NFC', self.cur.text)
        if key in ("_env", "_interpolate", "_meta") and not self.is_nole: self.error(f"Reserved key {key}")
        if key in obj and key != "<<": self.error(f"Duplicate key: {key}")
        self.next()
        if self.cur.type != Tok.Colon: self.error("Expected :")
        self.next(); val = self.parse_value()
        if key == "<<":
            if "<<" not in obj: obj["<<"] = []
            obj["<<"].append(val)
        else: obj[key] = val
    def parse_value(self) -> Any:
        self.depth += 1
        if self.depth > 100: self.error("Nesting limit exceeded")
        res = None
        if self.cur.type == Tok.Ampersand:
            self.next(); name = self.cur.text; self.next(); res = self.parse_value(); self.anchors[name] = res
        elif self.cur.type == Tok.Star:
            self.next(); res = "*" + self.cur.text; self.next()
        elif self.cur.type == Tok.LArrow:
            self.next(); t = self.cur.text; self.next(); self.next(); res = {"_coerce": {"type": t, "value": self.parse_value()}}
        elif self.cur.type == Tok.String: res = self.cur.text; self.next()
        elif self.cur.type == Tok.Number:
            s = self.cur.text; self.next()
            try:
                if '.' in s or 'e' in s.lower(): res = float(s)
                else: res = int(s)
            except ValueError: self.error(f"Invalid number: {s}")
        elif self.cur.type in (Tok.Date, Tok.True_, Tok.False_, Tok.Null):
            m = {Tok.Date: lambda x: x, Tok.True_: lambda x: True, Tok.False_: lambda x: False, Tok.Null: lambda x: None}
            res = m[self.cur.type](self.cur.text); self.next()
        elif self.cur.type == Tok.LBracket: res = self.parse_array()
        elif self.cur.type == Tok.LBrace: res = self.parse_object()
        else: self.error(f"Invalid value: {self.cur.type}")
        self.depth -= 1; return res
    def parse_array(self) -> List[Any]:
        self.next(); res = []
        while self.cur.type != Tok.RBracket:
            self.skip_newlines()
            if self.cur.type == Tok.RBracket: break
            res.append(self.parse_value()); self.skip_newlines()
            if self.cur.type == Tok.Comma: self.next()
        self.next()
        if len(res) > 100000: self.error("Array size limit")
        return res
    def parse_object(self) -> Dict[str, Any]:
        self.next(); res = {}
        while self.cur.type != Tok.RBrace:
            self.skip_newlines()
            if self.cur.type == Tok.RBrace: break
            self.parse_pair(res); self.skip_newlines()
            if self.cur.type == Tok.Comma: self.next()
        self.next()
        if len(res) > 10000: self.error("Object keys limit")
        return res
    def resolve_merges(self, val, visited=None):
        if visited is None: visited = []
        if id(val) in visited: raise NolError("Cycle detected")
        if isinstance(val, dict):
            visited.append(id(val))
            if "<<" in val:
                merges = val.pop("<<")
                for m in merges:
                    resolved_m = self.resolve_merges(m, visited)
                    if isinstance(resolved_m, str) and resolved_m.startswith("*"):
                        resolved_m = self.resolve_merges(self.anchors.get(resolved_m[1:]), visited)
                    if not isinstance(resolved_m, dict): raise NolError("Can only merge objects")
                    for k, v in resolved_m.items():
                        if k not in val: val[k] = v
            for k in list(val.keys()): val[k] = self.resolve_merges(val[k], visited)
            visited.pop()
        elif isinstance(val, list):
            visited.append(id(val))
            for i in range(len(val)): val[i] = self.resolve_merges(val[i], visited)
            visited.pop()
        elif isinstance(val, str) and val.startswith("*"):
            a = val[1:]
            if a not in self.anchors: raise NolError(f"Undefined anchor: {a}")
            return self.resolve_merges(self.anchors[a], visited)
        return val
    def resolve_env(self, val):
        if isinstance(val, dict):
            if "env" in val and len(val) == 1 and isinstance(val["env"], str):
                var = val["env"]
                if var in self.app_env_allowlist and var in self.doc_env_allowlist: return os.environ.get(var, "")
                raise NolError(f"Env access denied: {var}")
            for k in list(val.keys()): val[k] = self.resolve_env(val[k])
        elif isinstance(val, list):
            for i in range(len(val)): val[i] = self.resolve_env(val[i])
        return val
    def resolve_interpolations(self, val):
        if isinstance(val, dict):
            for k in list(val.keys()): val[k] = self.resolve_interpolations(val[k])
        elif isinstance(val, list):
            for i in range(len(val)): val[i] = self.resolve_interpolations(val[i])
        elif isinstance(val, str): return self.resolve_interpolation(val)
        return val
    def resolve_interpolation(self, s, depth=0):
        if depth > 50: raise NolError("Interpolation depth")
        def repl(match):
            parts = match.group(1).split('.'); curr = self.root
            for p in parts:
                if not isinstance(curr, dict) or p not in curr: raise NolError(f"Undefined: {match.group(1)}")
                curr = curr[p]
            if not isinstance(curr, (str, int, float, bool)): raise NolError("Not scalar")
            res = str(curr)
            return self.resolve_interpolation(res, depth+1) if "${" in res else res
        res = re.sub(r'\$\{([^}]+)\}', repl, s)
        if len(res) > 10 * 1024 * 1024: raise NolError("Interpolation size")
        return res
    def resolve_coercions(self, val):
        if isinstance(val, dict):
            if "_coerce" in val and len(val) == 1:
                c = val["_coerce"]; v = self.resolve_coercions(c["value"]); t = c["type"]
                if t == "int": return int(v)
                if t == "float": return float(v)
                if t == "bool":
                    if isinstance(v, bool): return v
                    if str(v).lower() == "true": return True
                    if str(v).lower() == "false": return False
                    raise NolError(f"Invalid bool coercion: {v}")
                if t == "string": return str(v)
                raise NolError(f"Unknown coercion: {t}")
            for k in list(val.keys()): val[k] = self.resolve_coercions(val[k])
        elif isinstance(val, list):
            for i in range(len(val)): val[i] = self.resolve_coercions(val[i])
        return val

class Document:
    def __init__(self, root): self.root = root
    def get(self, path=None):
        if path is None: return self.root
        curr = self.root
        for p in path.split('.'):
            if not isinstance(curr, dict) or p not in curr: return None
            curr = curr[p]
        return curr
    @staticmethod
    def parse(s, n=False, e=None): return Document(Parser(s, n, e).parse())

def parse(s, n=False, e=None): return Parser(s, n, e).parse()

File: libs/nol_nole.js
const fs = require('fs');
class NolError extends Error {
    constructor(m, l, c) { super(`${m}${l ? ` at ${l}:${c}` : ''}`); this.name = 'NolError'; this.line = l; this.col = c; }
}
const Tok = {
    EOF: 'EOF', Identifier: 'Identifier', String: 'String', Number: 'Number', Date: 'Date',
    True: 'True', False: 'False', Null: 'Null', Colon: 'Colon', Comma: 'Comma',
    LBracket: 'LBracket', RBracket: 'RBracket', LBrace: 'LBrace', RBrace: 'RBrace', Newline: 'Newline',
    Dot: 'Dot', Star: 'Star', LShift: 'LShift', Ampersand: 'Ampersand', LArrow: 'LArrow', RArrow: 'RArrow'
};
class Lexer {
    constructor(input) {
        if (input.length > 10 * 1024 * 1024) throw new NolError("Document size limit exceeded");
        if (input.includes('\0')) throw new NolError("Null bytes are prohibited");
        for (let i = 0; i < input.length; i++) { const cp = input.charCodeAt(i); if (cp >= 0xD800 && cp <= 0xDFFF) throw new NolError("Surrogate code points are prohibited"); }
        this.input = input; this.pos = 0; this.line = 1; this.col = 1;
    }
    peek(n = 0) { return this.pos + n < this.input.length ? this.input[this.pos + n] : '\0'; }
    advance() { const c = this.peek(); this.pos++; if (c === '\n') { this.line++; this.col = 1; } else this.col++; return c; }
    nextToken() {
        while (true) {
            const c = this.peek(); if (c === '\0') return { type: Tok.EOF, text: '', line: this.line, col: this.col };
            if (/\s/.test(c)) { if (c === '\n') { const t = { type: Tok.Newline, text: '\n', line: this.line, col: this.col }; this.advance(); return t; } this.advance(); continue; }
            if (c === '#') { this.advance(); if (this.peek() === '#') { this.advance(); while (this.peek() !== '\0') { if (this.peek() === '#' && this.peek(1) === '#') { this.advance(); this.advance(); break; } this.advance(); } } else { while (this.peek() !== '\0' && this.peek() !== '\n') this.advance(); } continue; }
            if (c === '"' || c === "'") return this.scanString();
            if (/\d/.test(c) || (c === '-' && /\d/.test(this.peek(1)))) return this.scanNumberOrDate();
            if (/[a-zA-Z_]/.test(c)) return this.scanIdentifier();
            const s = { ':': Tok.Colon, ',': Tok.Comma, '[': Tok.LBracket, ']': Tok.RBracket, '{': Tok.LBrace, '}': Tok.RBrace, '.': Tok.Dot, '*': Tok.Star, '&': Tok.Ampersand, '>': Tok.RArrow }[c];
            if (s) { const t = { type: s, text: c, line: this.line, col: this.col }; this.advance(); return t; }
            if (c === '<') { const l = this.line, col = this.col; this.advance(); if (this.peek() === '<') { this.advance(); return { type: Tok.LShift, text: '<<', line: l, col }; } return { type: Tok.LArrow, text: '<', line: l, col }; }
            throw new NolError(`Unexpected character: ${c}`, this.line, this.col);
        }
    }
    scanString() {
        const q = this.advance(); const l = this.line, col = this.col; const triple = this.peek() === q && this.peek(1) === q; if (triple) { this.advance(); this.advance(); }
        let res = '';
        while (true) {
            const c = this.peek(); if (c === '\0') throw new NolError("Unterminated string", l, col);
            if (triple) { if (c === q && this.peek(1) === q && this.peek(2) === q) { this.advance(); this.advance(); this.advance(); break; } }
            else { if (c === q) { this.advance(); break; } if (c === '\n') throw new NolError("Newline in string", l, col); }
            if (c === '\\' && q === '"') {
                this.advance(); const e = this.advance();
                switch (e) {
                    case '"': res += '"'; break; case "'": res += "'"; break; case '\\': res += '\\'; break;
                    case 'n': res += '\n'; break; case 'r': res += '\r'; break; case 't': res += '\t'; break;
                    case 'u': case 'U': { const n = e === 'u' ? 4 : 8; let u = ''; for (let i = 0; i < n; i++) u += this.advance(); const cp = parseInt(u, 16); if (cp === 0) throw new NolError("Null prohibited", l, col); if (cp >= 0xD800 && cp <= 0xDFFF) throw new NolError("Surrogate prohibited", l, col); res += String.fromCodePoint(cp); break; }
                    default: throw new NolError(`Invalid escape: \\${e}`, l, col);
                }
            } else res += this.advance();
            if (res.length > 1024 * 1024) throw new NolError("String length limit exceeded");
        }
        return { type: Tok.String, text: res, line: l, col };
    }
    scanNumberOrDate() {
        const l = this.line, c = this.col; let s = ''; while (/[a-zA-Z0-9\-\._:T\+]/.test(this.peek())) s += this.advance();
        if (/^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(Z|[+-]\d{2}:\d{2})?)?$/.test(s)) return { type: Tok.Date, text: s, line: l, col: c };
        return { type: Tok.Number, text: s, line: l, col: c };
    }
    scanIdentifier() { const l = this.line, c = this.col; let s = ''; while (/[a-zA-Z0-9_-]/.test(this.peek())) s += this.advance(); const kw = { 'true': Tok.True, 'false': Tok.False, 'null': Tok.Null }; return { type: kw[s] || Tok.Identifier, text: s, line: l, col: c }; }
}
class Parser {
    constructor(i, n = false, e = []) { this.lexer = new Lexer(i); this.cur = this.lexer.nextToken(); this.isNole = n; this.appEnvAllowlist = new Set(e); this.docEnvAllowlist = new Set(); this.anchors = new Map(); this.root = {}; this.depth = 0; this.startTime = Date.now(); }
    error(m) { throw new NolError(m, this.cur.line, this.cur.col); }
    next() { this.cur = this.lexer.nextToken(); }
    skipNewlines() { while (this.cur.type === Tok.Newline) this.next(); }
    parse() {
        while (this.cur.type !== Tok.EOF) { this.skipNewlines(); if (this.cur.type === Tok.EOF) break; if (this.cur.type === Tok.LBracket) this.parseSection(); else if (this.isKeyToken()) this.parsePair(this.root); else this.error(`Unexpected token: ${this.cur.type}`); if (Date.now() - this.startTime > 1000) this.error("Parse timeout"); }
        if (this.isNole) { this.collectEnvAllowlist(this.root); this.root = this.resolveMerges(this.root); this.root = this.resolveEnv(this.root); this.root = this.resolveInterpolations(this.root); this.root = this.resolveCoercions(this.root); }
        return this.root;
    }
    collectEnvAllowlist(o) { if (o && typeof o === 'object') { if (o._env && o._env.allowed && Array.isArray(o._env.allowed)) o._env.allowed.forEach(a => this.docEnvAllowlist.add(String(a))); Object.values(o).forEach(v => this.collectEnvAllowlist(v)); } }
    isKeyToken() { return [Tok.Identifier, Tok.String, Tok.LShift].includes(this.cur.type); }
    parseSection() {
        this.next(); let path = [], isArray = false; while (true) { if (this.cur.type === Tok.Star) { isArray = true; this.next(); } else if ([Tok.Identifier, Tok.String].includes(this.cur.type)) { path.push(this.cur.text.normalize('NFC')); this.next(); } else this.error("Expected section name"); if (this.cur.type === Tok.Dot) { this.next(); continue; } if (this.cur.type === Tok.RBracket) { this.next(); break; } this.error("Expected . or ]"); }
        let target = this.root;
        for (let i = 0; i < path.length; i++) { const part = path[i]; if (part in target) { if (typeof target[part] !== 'object' && !(isArray && i === path.length - 1 && Array.isArray(target[part]))) this.error(`Collision at ${part}`); } else target[part] = (isArray && i === path.length - 1) ? [] : {}; if (!(isArray && i === path.length - 1)) target = target[part]; }
        if (isArray) { const arr = target[path[path.length - 1]], newObj = {}; arr.push(newObj); target = newObj; }
        this.skipNewlines(); while (this.isKeyToken()) { this.parsePair(target); this.skipNewlines(); }
    }
    parsePair(obj) { const k = this.cur.type === Tok.LShift ? '<<' : this.cur.text.normalize('NFC'); if (['_env', '_interpolate', '_meta'].includes(k) && !this.isNole) this.error(`Reserved key ${k}`); if (k in obj && k !== '<<') this.error(`Duplicate key: ${k}`); this.next(); if (this.cur.type !== Tok.Colon) this.error("Expected :"); this.next(); const val = this.parseValue(); if (k === '<<') { if (!obj['<<']) obj['<<'] = []; obj['<<'].push(val); } else obj[k] = val; }
    parseValue() {
        this.depth++; if (this.depth > 100) this.error("Nesting limit exceeded");
        let res;
        switch (this.cur.type) {
            case Tok.Ampersand: this.next(); const n = this.cur.text; this.next(); res = this.parseValue(); this.anchors.set(n, res); break;
            case Tok.Star: this.next(); res = '*' + this.cur.text; this.next(); break;
            case Tok.LArrow: this.next(); const t = this.cur.text; this.next(); if (this.cur.type !== Tok.RArrow) this.error("Expected >"); this.next(); res = { _coerce: { type: t, value: this.parseValue() } }; break;
            case Tok.String: res = this.cur.text; this.next(); break;
            case Tok.Number: const s = this.cur.text; this.next(); res = (s.includes('.') || s.toLowerCase().includes('e')) ? parseFloat(s) : parseInt(s, 10); break;
            case Tok.True: res = true; this.next(); break; case Tok.False: res = false; this.next(); break; case Tok.Null: res = null; this.next(); break;
            case Tok.Date: res = this.cur.text; this.next(); break;
            case Tok.LBracket: res = this.parseArray(); break; case Tok.LBrace: res = this.parseObject(); break;
            default: this.error(`Invalid value: ${this.cur.type}`);
        }
        this.depth--; return res;
    }
    parseArray() { this.next(); const res = []; while (this.cur.type !== Tok.RBracket) { this.skipNewlines(); if (this.cur.type === Tok.RBracket) break; res.push(this.parseValue()); this.skipNewlines(); if (this.cur.type === Tok.Comma) this.next(); } this.next(); if (res.length > 100000) this.error("Array size limit"); return res; }
    parseObject() { this.next(); const res = {}; while (this.cur.type !== Tok.RBrace) { this.skipNewlines(); if (this.cur.type === Tok.RBrace) break; this.parsePair(res); this.skipNewlines(); if (this.cur.type === Tok.Comma) this.next(); } this.next(); if (Object.keys(res).length > 10000) this.error("Object keys limit"); return res; }
    resolveMerges(val, visited = new Set()) {
        if (visited.has(val)) throw new NolError("Cycle detected");
        if (val && typeof val === 'object' && !Array.isArray(val)) {
            visited.add(val); if (val['<<']) { const merges = val['<<']; delete val['<<']; for (let m of merges) { let rm = this.resolveMerges(m, visited); if (typeof rm === 'string' && rm.startsWith('*')) rm = this.resolveMerges(this.anchors.get(rm.substring(1)), visited); if (!rm || typeof rm !== 'object' || Array.isArray(rm)) throw new NolError("Can only merge objects"); for (let [k, v] of Object.entries(rm)) if (!(k in val)) val[k] = v; } }
            for (let k of Object.keys(val)) val[k] = this.resolveMerges(val[k], visited); visited.delete(val);
        } else if (Array.isArray(val)) { visited.add(val); for (let i = 0; i < val.length; i++) val[i] = this.resolveMerges(val[i], visited); visited.delete(val); }
        else if (typeof val === 'string' && val.startsWith('*')) { const a = val.substring(1); if (!this.anchors.has(a)) throw new NolError(`Undefined anchor: ${a}`); return this.resolveMerges(this.anchors.get(a), visited); }
        return val;
    }
    resolveEnv(v) {
        if (v && typeof v === 'object' && !Array.isArray(v)) { if (v.env && Object.keys(v).length === 1 && typeof v.env === 'string') { if (this.appEnvAllowlist.has(v.env) && this.docEnvAllowlist.has(v.env)) return process.env[v.env] || ''; throw new NolError(`Env denied: ${v.env}`); } for (let k of Object.keys(v)) v[k] = this.resolveEnv(v[k]); }
        else if (Array.isArray(v)) { for (let i = 0; i < v.length; i++) v[i] = this.resolveEnv(v[i]); } return v;
    }
    resolveInterpolations(v) { if (v && typeof v === 'object' && !Array.isArray(v)) { for (let k of Object.keys(v)) v[k] = this.resolveInterpolations(v[k]); } else if (Array.isArray(v)) { for (let i = 0; i < v.length; i++) v[i] = this.resolveInterpolations(v[i]); } else if (typeof v === 'string') return this.resolveInterpolation(v); return v; }
    resolveInterpolation(s, depth = 0) {
        if (depth > 50) throw new NolError("Interpolation depth limit");
        const res = s.replace(/\$\{([^}]+)\}/g, (_, path) => { const parts = path.split('.'); let curr = this.root; for (let p of parts) { if (!curr || typeof curr !== 'object' || !(p in curr)) throw new NolError(`Undefined: ${path}`); curr = curr[p]; } if (!['string', 'number', 'boolean'].includes(typeof curr) && curr !== null) throw new NolError("Not scalar"); const vs = String(curr); return vs.includes('${') ? this.resolveInterpolation(vs, depth + 1) : vs; });
        if (res.length > 10 * 1024 * 1024) throw new NolError("Interpolation size limit"); return res;
    }
    resolveCoercions(v) { if (v && typeof v === 'object' && !Array.isArray(v)) { if (v._coerce && Object.keys(v).length === 1) { const { type: t, value: rv } = v._coerce; const val = this.resolveCoercions(rv); switch (t) { case 'int': return parseInt(val, 10); case 'float': return parseFloat(val); case 'bool': return typeof val === 'boolean' ? val : (String(val).toLowerCase() === 'true'); case 'string': return String(val); default: throw new NolError(`Unknown coercion: ${t}`); } } for (let k of Object.keys(v)) v[k] = this.resolveCoercions(v[k]); } else if (Array.isArray(v)) { for (let i = 0; i < v.length; i++) v[i] = this.resolveCoercions(v[i]); } return v; }
}
class Document {
    constructor(r) { this.root = r; }
    get(p) { if (!p) return this.root; let curr = this.root; for (let part of p.split('.')) { if (!curr || typeof curr !== 'object' || !(part in curr)) return undefined; curr = curr[part]; } return curr; }
    static parse(s, n = false, e = []) { return new Document(new Parser(s, n, e).parse()); }
}
function parse(i, n = false, e = []) { return new Parser(i, n, e).parse(); }
module.exports = { parse, NolError, Document };
